watertap/core/zero_order_sido_reactive.py

```diff
@@ -0,0 +1,416 @@
+                    cratio = -1
+                else:
+                    nu_j = cratio = stoich[j]["order"]
+                    try:
+                        nu_k = cratio = stoich[key_reactant]["order"]
+                    except KeyError:
+                        raise KeyError(
+                            f"{self.name} - database provided does not "
+                            f"contain an entry for order w.r.t. species "
+                            f"{key_reactant} in reaction {r}.")
+                    try:
+                        MW_j = cratio = stoich[j]["molecular_weight"]
+                    except KeyError:
+                        raise KeyError(
+                            f"{self.name} - database provided does not "
+                            f"contain an entry for molecular_weight w.r.t. "
+                            f"species {j} in reaction {r}.")
+                    try:
+                        MW_k = cratio = stoich[key_reactant]["molecular_weight"]
+                    except KeyError:
+                        raise KeyError(
+                            f"{self.name} - database provided does not "
+                            f"contain an entry for molecular_weight w.r.t. "
+                            f"species {key_reactant} in reaction {r}.")
+                    cratio = -(nu_j*MW_j)/(nu_k*MW_k)
+            else:
+                raise RuntimeError(
+                    f"{self.name} - database provided does not "
+                    f"contain any information for conversion_ratio or reaction "
+                    f"order w.r.t. species {j} in reaction {r}.")
+            p.set_value(cratio)
+
+    # Add performance constraints
+    # Water recovery
+    @self.Constraint(self.flowsheet().time, doc='Water recovery equation')
+    def water_recovery_equation(b, t):
+        return (b.recovery_frac_mass_H2O[t] *
+                (b.properties_in[t].flow_mass_comp["H2O"] +
+                 sum(b.generation_ratio[r, "H2O"]*b.extent_of_reaction[t, r]
+                     for r in self.reaction_set)) ==
+                b.properties_treated[t].flow_mass_comp["H2O"])
+
+    # Flow balance
+    @self.Constraint(self.flowsheet().time, doc='Overall flow balance')
+    def water_balance(b, t):
+        return (b.properties_in[t].flow_mass_comp["H2O"] +
+                sum(b.generation_ratio[r, "H2O"]*b.extent_of_reaction[t, r]
+                    for r in self.reaction_set) ==
+                b.properties_treated[t].flow_mass_comp["H2O"] +
+                b.properties_byproduct[t].flow_mass_comp["H2O"])
+
+    # Solute removal
+    @self.Constraint(self.flowsheet().time,
+                     self.config.property_package.solute_set,
+                     doc='Solute removal equations')
+    def solute_removal_equation(b, t, j):
+        return (b.removal_frac_mass_solute[t, j] *
+                (b.properties_in[t].flow_mass_comp[j] +
+                 sum(b.generation_ratio[r, j]*b.extent_of_reaction[t, r]
```

Comment:
I think it would be helpful to the users if we actually made a generation term (i.e. indexed variable and constraint) rather than just using the product of the generation_ratio and extent of reaction. I could see users frequently wanting to just check how much of a component was generated.